### ISSUE TITLE
Recipe improvements

### ISSUE THREAD

**Issue by h-vetinari** [conda-forge/torchvision-feedstock#22]
Some open points remaining from (OP of) #12:

* [x] build with cuda
* [ ] build with ffmpeg support
* [ ] turn into multi-output recipe (e.g. with/without ffmpeg/image support)?
* [ ] package c++ lib separately?
* [x] build for osx
* [ ] build for win

CC @hmaarrfk @isuruf

**Comment by hmaarrfk** [comment]
Thoughts:

- build with ffmpeg support
Sure

- turn into multi-output recipe (e.g. with/without ffmpeg/image support)?
Yes

- package c++ lib separately?
Maybe just create a split package and have the `lib` package not depend on python https://github.com/conda-forge/opencv-feedstock/blob/master/recipe/meta.yaml#L1

- build for osx
Thanks!!!!

- build for win

Didn't know this was dropped but good to keep track!

**Comment by h-vetinari** [comment]
> > package c++ lib separately?
>
> Maybe just create a split package and have the `lib` package not depend on python https://github.com/conda-forge/opencv-feedstock/blob/master/recipe/meta.yaml#L1

That's (another aspect of) what I meant by making the recipe have multiple outputs

> > build for osx
>
> Thanks!!!!

That credit belongs to @isuruf

> > build for win
>
> Didn't know this was dropped but good to keep track!

When I picked up the feedstock again, linux [was](https://github.com/conda-forge/torchvision-feedstock/pull/12/files#r595572323) the only build variant... Not sure how much effort this would be, but the pytorch channel has windows builds (even with cuda).

**Comment by hmaarrfk** [comment]
Also cross referencing this issue:
https://github.com/conda-forge/pytorch-cpu-feedstock/issues/40#issuecomment-800611656

**Comment by Tobias-Fischer** [comment]
It would also be great if the correct version of pytorch is being pinned. Currently there can be a mismatch between the pytorch and torchvision versions. The compatibility table is shown in [their GitHub repo ](https://github.com/pytorch/vision)

**Comment by hmaarrfk** [comment]
As test files are being copied into the recipe archive, the many binary files are actually increasing the size of said archive:
![image](https://user-images.githubusercontent.com/90008/150683074-fc130a16-8cba-4664-9270-74eec792099c.png)

**Comment by mgorny** [comment]
> turn into multi-output recipe (e.g. with/without ffmpeg/image support)?

I'd like to give this a shot, however I'm wondering if I should go for rattler-build (like torchaudio has) before doing that, i.e. while the conversion is still rather easy, and the new complexity would be done directly on rattler-build? Or do you prefer to stay with conda-build on this recipe for now?

**Comment by hmaarrfk** [comment]
>     turn into multi-output recipe (e.g. with/without ffmpeg/image support)?

The problem I have with this, is that I believe that ffmpeg support, specifically libav filter support is fundamentally broken in torchvision.

Last I checked, they were pinned on FFMPEG4 https://github.com/pytorch/vision/blob/d3beb52a00e16c71e821e192bcc592d614a490c0/packaging/pre_build_script.sh#L20
which is way old with many deprecations that got into effect in FFMPEG 5/6/7


https://github.com/PyAV-Org/PyAV/pull/1472 shows some of the challenges I am aware of in the 6->7 transition.

Furthermore, the main goal in my mind of using FFMPEG + torchvision was that you would be able to pipe a video encoded in one of Nvidia's decoders, directly into some ML model. When I tried to do this, I got a bunch of segfaults.
The problems are "worse" when checks for "safe" downgrades to a naive path are made poorly, as is often done in unmaintained research grade codepaths.

TLDR: I'm ok with enabling FFMPEG, and work toward splitting the package, but somebody knowledgeable in the usage will have to show me it isn't segfaulting in 2025.

**Comment by hmaarrfk** [comment]
In 2022, I had flagged "TestVideoReader" as a problematic test.

seems very suspicious!!!  https://github.com/conda-forge/torchvision-feedstock/pull/60/files#diff-f3725a55bf339595bf865fec73bda8ac99f283b0810c205442021f29c06eea9aR148

**Comment by hmaarrfk** [comment]
I just unchecked the “build with ffmpeg support” since I believe we disabled it recently

**Comment by mgorny** [comment]
Okay, I guess rattler-build won't fly because of the `pytorch ={{ compatible_pytorch }}.*={{ torch_proc_type }}*` syntax.

**Comment by Tobias-Fischer** [comment]
> Okay, I guess rattler-build won't fly because of the `pytorch ={{ compatible_pytorch }}.*={{ torch_proc_type }}*` syntax.

/cc @wolfv  for visibility - do you have plans to support something like this?

**Comment by wolfv** [comment]
If you remove the = signs it works the same and should be less confusing :)

**Comment by wolfv** [comment]
You can either use a syntax like "foo 1.2.* foobar" or "foo 1.2.* [build=foobar]" if you want to be more explicit

**Comment by h-vetinari** [comment]
> Okay, I guess rattler-build won't fly because of the `pytorch ={{ compatible_pytorch }}.*={{ torch_proc_type }}*` syntax.

I'm not going to block a conversion to rattler over this topic. Please use the notation with the square brackets that Wolf suggested

**Comment by mgorny** [comment]
> I'm not going to block a conversion to rattler over this topic. Please use the notation with the square brackets that Wolf suggested

Ah, okay, thanks!

**Comment by mgorny** [comment]
Ok, so after some initial looking, I don't think there's much of a point in splitting the image library. We could technically provide different variants of it (i.e. with different formats supported), but in the end I think the dependencies are relatively small here.

FWICS we currently don't enable HEIF/AVIF image support. At least the latter is getting popular lately, so it'd probably make sense to enable it. Along with its dependencies it is ~20M of packages, so that might justify making it optional. However, it's still tiny compared to PyTorch itself.

I'm going to look into how bad things look wrt ffmpeg — if we were to enable that, it would probably make sense to make that into a non-default configuration (if only because it would mean pinning to an old version).

**Comment by mgorny** [comment]
Ok, scratch that about ffmpeg. The versions matching upstream pin require old `zlib` that's incompatible with `libprotobuf` used by `pytorch`.

**Comment by hmaarrfk** [comment]
> FWICS we currently don't enable HEIF/AVIF image support.

what would be required for this? can you point me to the build option?

**Comment by mgorny** [comment]
> > FWICS we currently don't enable HEIF/AVIF image support.
> 
> what would be required for this? can you point me to the build option?

`TORCHVISION_USE_HEIC` and `TORCHVISION_USE_AVIF` — both have a `# TODO enable by default!` :-).

https://github.com/pytorch/vision/blob/3ac97aa9120137381ed1060f37237e44485ac2aa/setup.py#L22-L23

I haven't tested to verify yet (particularly if the newest versions are compatible), but the dependencies are probably `libheif` and `libavif` respectively. On my `base` environment, they pull in:

libheif:
```
  aom                conda-forge/linux-64::aom-3.9.1-hac33072_0 
  dav1d              conda-forge/linux-64::dav1d-1.2.1-hd590300_0 
  libavif16          conda-forge/linux-64::libavif16-1.1.1-h1909e37_2 
  libde265           conda-forge/linux-64::libde265-1.0.15-h00ab1b0_0 
  libheif            conda-forge/linux-64::libheif-1.18.2-gpl_hffcb242_100 
  rav1e              conda-forge/linux-64::rav1e-0.6.6-he8a937b_2 
  svt-av1            conda-forge/linux-64::svt-av1-2.3.0-h5888daf_0 
  x265               conda-forge/linux-64::x265-3.5-h924138e_3 
```

libavif:
```
  aom                conda-forge/linux-64::aom-3.9.1-hac33072_0 
  dav1d              conda-forge/linux-64::dav1d-1.2.1-hd590300_0 
  libavif            conda-forge/linux-64::libavif-1.1.1-h1909e37_2 
  libavif16          conda-forge/linux-64::libavif16-1.1.1-h1909e37_2 
  rav1e              conda-forge/linux-64::rav1e-0.6.6-he8a937b_2 
  svt-av1            conda-forge/linux-64::svt-av1-2.3.0-h5888daf_0 
```

**Comment by hmaarrfk** [comment]
Nice, do you want to attempt heif/libavif in https://github.com/conda-forge/torchvision-feedstock/pull/108

I would ask that you add a dependency on `libheif =* lgpl*` at build time.

**Comment by mgorny** [comment]
Sure. Yeah, noticed `lgpl*` there and had the same idea.

**Comment by roaldarbol** [comment]
With `pytorch` now building on Windows (yay!), would it be possible to get torchvision Windows builds too?

**Comment by h-vetinari** [comment]
#111

**Comment by roaldarbol** [comment]
Ah, brilliant, hadn't seen that! Great to see! 😃
